src/batched/KokkosBatched_Gemm_Handle.hpp

```diff
@@ -0,0 +1,178 @@
+}
+
+/// \brief KokkosBatched algorithm types. See BatchedGemmHandle for details.
+namespace GemmKokkosBatchedAlgos {
+enum GEMM_KOKKOS_BATCHED_ALGOS : int {
+  KK_TEAM = GemmTplAlgos::N,
+  KK_TEAMVECTOR,
+  KK_SERIALSIMD,
+  KK_TEAMSIMD,
+  KK_SERIAL_OPT2,
+  KK_SERIAL_SHMEM,
+  KK_DBLBUF,
+  N
+};
+}
+
+// clang-format off
+/// \brief Handle for selecting runtime behavior of the BatchedGemm interface.
+///
+/// \var kernelAlgoType  Specifies which algorithm to use for invocation (default, SQUARE).
+///
+///                    Specifies whether to select optimal invocations based on inputs and
+///                    heuristics:
+///                      SQUARE select invocations based on square matrix heuristics where M=N
+///                      TALL   select invocations based on tall   matrix heuristics where M>N
+///                      WIDE   select invocations based on wide   matrix heuristics where M<N
+///                    Note: If the heuristics indicate SIMD views are required for optimal
+///                    performance, notify the user that SIMD views are required for
+///                    optimal performance.
+///
+///                    Specifies which cmake-enabled tpl algorithm to invoke:
+///                      ARMPL    Invoke the ArmPL TPL interface
+///                      MKL      Invoke the MKL TPL interface
+///                      CUBLAS   Invoke the CuBLAS TPL interface
+///                      MAGMA    Invoke the Magma TPL interface
+///                    Note: Requires that input views for A, B, and C reside on either host
+///                    or device depending on the TPL selected.
+///                    Note: If the user selects a TPL, an error will be thrown if:
+///                       1. The TPL is not enabled via cmake
+///                       2. The input views do not reside on the host/device as needed
+///
+///                    Specifies which kokkos-kernels (KK) algorithm to invoke:
+///                      KK_SERIAL       Invoke SerialGemm     via RangePolicy(BatchSz)
+///                      KK_TEAM         Invoke TeamGemm       via TeamPolicy(BatchSz)
+///                      KK_TEAMVECTOR   Invoke TeamVectorGemm via TeamPolicy(BatchSz)
+///                      KK_SERIALSIMD   Invoke SerialGemm     via TeamPolicy(BatchSz)
+///                      KK_TEAMSIMD     Invoke TeamGemm       via TeamPolicy(BatchSz)
+///                      KK_SERIAL_OPT2  Invoke SerialGemm     via RangePolicy(BatchSz*N*M)
+///                      KK_SERIAL_SHMEM Invoke SerialGemm     via TeamPolicy(BatchSz)
+///                                      Copies A and B to shared memory before GEMM.
+///                      KK_DBLBUF       Solve GEMM            via TeamPolicy(BatchSz*TILES)
+///                                      Uses custom functor, tiling and double buffering
+///                                      via shared memory and register buffers.
```

Comment:
@ajpowelsnl, @cwpearson: We would like to get your feedback on the clarity of these `KK_` algorithm names. From a user perspective, are these names confusing? If so, would it help to include the execution policy (`RangePolicy` or `TeamPolicy`) in the `KK_` algorithm name? Thank you in advance for your input.

Reply:
I'm most confused about `SERIAL_OPT2` and `SERIAL_SHMEM`

If `BatchSz` is the number of GEMMs we're doing, here's what I think these mean:
* `KK_SERIAL` - each thread does a GEMM
* `KK_TEAM` - each team does a GEMM
* `KK_TEAMVECTOR` - each team, with hierarchical parallelism somehow (maybe for a dot product?)
* `KK_SERIALSIMD`/`KK_TEAMSIMD` I assume this is serial/team but with some explicit SIMD parts
* `KK_SERIALOPT2` a mysterious alternative to `KK_SERIAL` (OPT suggests it may be faster?)
* `KK_SERIALSHMEM`: Somehow a team must be collaborating, otherwise I'm not sure how helpful shared memory is? this one is mysterious too.
* KK_DBLBUF: the best version (why wouldn't I use this one?)

Reply:
@cwpearson, does https://github.com/kokkos/kokkos-kernels/pull/1054/commits/72dd082a1e18449b6617bee26ed257a2e8d59364 clear things up?

Reply:
Looks good to me!